### PR TITLE
Provides new Loadbalancer implementation

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR introduced a reworked version of the loadbalancer.

This one should be a final draft

### CODE REVIEW

File: rsocket-core/src/main/java/io/rsocket/loadbalance/LoadBalancedRSocketClient.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class LoadBalancedRSocketClient implements RSocketClient {
+
+  private final RSocketPool rSocketPool;
+
+  LoadBalancedRSocketClient(RSocketPool rSocketPool) {
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/loadbalance/LoadBalancedRSocketClient.java:30]:
```suggestion
  private LoadBalancedRSocketClient(RSocketPool rSocketPool) {
```

File: rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+public interface PooledRSocket extends RSocket {
+
+  Stats stats();
+
+  LoadbalanceTarget supplier();
+
+  boolean markForRemoval();
+
+  boolean markActive();
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java:28]: Add some Javadoc on semantics, what happens or should happen, return values, etc. There is some Javadoc in the  concrete class `ResolvingPooledRSocket`. That could be moved up here and used as starting point.

File: rsocket-core/src/main/java/io/rsocket/loadbalance/ResolvingPooledRSocket.java [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Operators;
+import reactor.util.context.Context;
+
+/** Specific interface for all RSocket store in {@link RSocketPool} */
+final class ResolvingPooledRSocket extends ResolvingOperator<RSocket>
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/loadbalance/ResolvingPooledRSocket.java:32]:
```suggestion
/** Default implementation of {@link PooledRSocket} stored in {@link RSocketPool} */
final class DefaultPooledRSocket extends ResolvingOperator<RSocket>
```

File: rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package io.rsocket.loadbalance;
+
+import io.rsocket.RSocket;
+
+public interface PooledRSocket extends RSocket {
+
+  Stats stats();
+
+  LoadbalanceTarget supplier();
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java]: fixed

File: rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java (hunk 2 above)
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/loadbalance/PooledRSocket.java]: Fixed

File: rsocket-core/src/main/java/io/rsocket/loadbalance/LoadbalanceTarget.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.loadbalance;
+
+import io.rsocket.RSocket;
+import reactor.core.publisher.Mono;
+
+public class LoadbalanceTarget {
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/loadbalance/LoadbalanceTarget.java]: fixed

File: rsocket-core/src/main/java/io/rsocket/loadbalance/Stats.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    this.errorStamp = now;
+    this.stamp0 = now;
+    this.duration = 0L;
+    this.pending = 0;
+    this.median = new Median();
+    this.interArrivalTime = new Ewma(1, TimeUnit.MINUTES, DEFAULT_INITIAL_INTER_ARRIVAL_TIME);
+    this.errorPercentage = new Ewma(5, TimeUnit.SECONDS, 1.0);
+    this.tau = Clock.unit().convert((long) (5 / Math.log(2)), TimeUnit.SECONDS);
+  }
+
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/loadbalance/Stats.java:66]: fixed

File: rsocket-core/src/main/java/io/rsocket/loadbalance/RSocketPool.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+  static final AtomicReferenceFieldUpdater<RSocketPool, Subscription> S =
+      AtomicReferenceFieldUpdater.newUpdater(RSocketPool.class, Subscription.class, "s");
+
+  RSocketPool(Publisher<List<LoadbalanceTarget>> source, LoadbalanceStrategy loadbalanceStrategy) {
+    this.loadbalanceStrategy = loadbalanceStrategy;
+    if (loadbalanceStrategy instanceof StatsBasedLoadbalanceStrategy) {
+      this.statsSupplier = Stats::new;
+    } else {
+      this.statsSupplier = () -> Stats.NoOpsStats.INSTANCE;
+    }
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/loadbalance/RSocketPool.java]: Fixed

**Pull request merged**